bliss/encoder/convnet.py

```diff
@@ -52,26 +52,81 @@ def forward(self, x):
 
 
 class FeaturesNet(nn.Module):
-    def __init__(self, n_bands, ch_per_band, num_features, double_downsample=True):
+    def __init__(self, n_bands, ch_per_band, num_features, tile_slen=2, downsample_at_front=False):
         super().__init__()
-
         nch_hidden = 64
         self.preprocess3d = nn.Sequential(
             nn.Conv3d(n_bands, nch_hidden, [ch_per_band, 5, 5], padding=[0, 2, 2]),
             nn.BatchNorm3d(nch_hidden),
             nn.SiLU(),
         )
-        self.backbone = nn.Sequential(
-            ConvBlock(nch_hidden, 64, kernel_size=5, padding=2),
-            nn.Sequential(*[ConvBlock(64, 64, kernel_size=5, padding=2) for _ in range(4)]),
-            ConvBlock(64, 128, stride=2),
-            nn.Sequential(*[ConvBlock(128, 128) for _ in range(5)]),
-            ConvBlock(128, num_features, stride=(2 if double_downsample else 1)),  # 4
-        )
+
+        # Currently, if tile length is 2 or 4,
+        # we revert to previous version of the network for backward compatibility
+        # This needs to be fixed in future iterations
+        # to make the network architecture cleaner and generalizable
+        # for tile sizes of 2 or 4, downsampling at front is disabled for now
+        # for backward compatibility, might be useful to add later
+
+        log_tile_size = torch.log2(torch.tensor(tile_slen))
+        num_downsample = int(torch.round(log_tile_size)) - 1
+        self.backbone = nn.ModuleList()
+        if tile_slen == 2:
+            self.backbone.append(ConvBlock(nch_hidden, 64, kernel_size=5, padding=2))
+            self.backbone.append(
+                nn.Sequential(*[ConvBlock(64, 64, kernel_size=5, padding=2) for _ in range(4)])
+            )
+            self.backbone.append(ConvBlock(64, 128, stride=2))
+            self.backbone.append(nn.Sequential(*[ConvBlock(128, 128) for _ in range(5)]))
+            self.backbone.append(ConvBlock(128, num_features, stride=1))
+        elif tile_slen == 4:
+            self.backbone.append(ConvBlock(nch_hidden, 64, kernel_size=5, padding=2))
+            self.backbone.append(
+                nn.Sequential(*[ConvBlock(64, 64, kernel_size=5, padding=2) for _ in range(4)])
+            )
+            self.backbone.append(ConvBlock(64, 128, stride=2))
+            self.backbone.append(nn.Sequential(*[ConvBlock(128, 128) for _ in range(5)]))
+            self.backbone.append(ConvBlock(128, num_features, stride=2))
+        else:
```

Comment:
@kapnadak @gapatron  Now that I'm see this, I think it'd be better if we left `convnet.py` unmodified for your case study. Instead, we can define a new version of `FeaturesNet` within your case study directory, which can be modified without affecting other case studies. PR #1003 should make it possible to swap in arbitrary CNN architectures by defining a subclass of `Encoder`.